### ISSUE TITLE
incorrect calculation of learning rate

### ISSUE THREAD

**Issue by lolpa1n** [ultralytics/ultralytics#20768]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

Hello, is this the correct work of changing the learning rate? 
My parameters 
batch size - 48,
lr0 - 0.0001, 
lrf - 0.1, 
warmap epochs - 5,
I expect that after 5 epochs there will be lr 0.0001 and by the end it will decrease by 10 times. However, the warm-up takes 20 epochs, why does it not take my value but takes it based on nbs? If I make 16 batch, then the warm-up is 5 epochs and the correct value is about 0.0001, and at 48 it only reaches 0.00007

![Image](https://github.com/user-attachments/assets/d49ffc16-0297-427f-89e5-2c16b815717b)

### Environment

Ultralytics 8.3.143 🚀 Python-3.10.11 torch-2.5.1+cu124 CUDA:0 (NVIDIA RTX 6000 Ada Generation, 48647MiB)
Setup complete ✅ (256 CPUs, 1007.7 GB RAM, 2423.4/3519.5 GB disk)

OS                  Linux-6.1.52-1.el7.3.x86_64-x86_64-with-glibc2.31
Environment         Linux
Python              3.10.11
Install             pip
Path                /opt/conda/lib/python3.10/site-packages/ultralytics
RAM                 1007.65 GB
Disk                2423.4/3519.5 GB
CPU                 AMD EPYC 7763 64-Core Processor
CPU count           256
GPU                 NVIDIA RTX 6000 Ada Generation, 48647MiB
GPU count           7
CUDA                12.4

numpy               ✅ 2.1.0>=1.23.0
matplotlib          ✅ 3.9.3>=3.3.0
opencv-python       ✅ 4.9.0.80>=4.6.0
pillow              ✅ 11.0.0>=7.1.2
pyyaml              ✅ 6.0>=5.3.1
requests            ✅ 2.31.0>=2.23.0
scipy               ✅ 1.14.1>=1.4.1
torch               ✅ 2.5.1>=1.8.0
torch               ✅ 2.5.1!=2.4.0,>=1.8.0; sys_platform == "win32"
torchvision         ✅ 0.20.1>=0.9.0
tqdm                ✅ 4.67.1>=4.64.0
psutil              ✅ 5.9.0
py-cpuinfo          ✅ 9.0.0
pandas              ✅ 2.2.3>=1.1.4
ultralytics-thop    ✅ 2.0.12>=2.0.0

### Minimal Reproducible Example

from ultralytics import YOLO

model = YOLO("yolo11s-seg.pt")

project = "synth-segm-pump"
name = "yolo11-m-640-orig-lr0001-50"

results = model.train(project=project, 
                      name=name, 
                      data="data.yaml", 
                      epochs=50,
                      imgsz=640,
                      device=0,
                      batch=48,
                      single_cls=True,
                      cos_lr=True,
                      warmup_epochs=5,
                      lr0= 0.0001,
                      lrf= 0.1,
                      optimizer= 'AdamW',
                      )

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @lolpa1n, thank you for bringing this up and for providing detailed information and a reproducible example! 🚀 This is an automated response to help streamline support. An Ultralytics engineer will review your issue and assist you further soon.

We suggest checking out the [Docs](https://docs.ultralytics.com/) for guidance on training configurations and learning rate scheduling, as well as [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples.

If this is a 🐛 Bug Report, thank you for including a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/)—this is very helpful for debugging.

If you have further details, such as additional training logs, screenshots, or dataset samples, please share them here to assist the team. For best practices during training, you can review our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best:
- For real-time chat: [Discord](https://discord.com/invite/ultralytics) 🎧
- For in-depth discussion: [Discourse](https://community.ultralytics.com/)
- Or join threads on our [Subreddit](https://reddit.com/r/Ultralytics)

## Upgrade

To ensure you’re not encountering an issue already fixed in the latest version, please upgrade the `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for your patience and for helping us improve Ultralytics!

**Comment by Y-T-G** [comment]
It runs for at least 100 batches. So if you have less data, then it will not follow epochs, but iterations.

https://github.com/ultralytics/ultralytics/blob/9584162f62043a250b461b42855b19834f68ad79/ultralytics/engine/trainer.py#L335
